src/parser/lexer.rs

```diff
@@ -0,0 +1,140 @@
+use std::{iter::Peekable, str::Chars};
+
+use super::token::Token;
+use super::range::{Range, Position};
+
+/// The lexer state
+#[derive(Debug)]
+pub struct Lexer<'a> {
+    pub input: &'a str,
+    pub position: usize,
+    pub peekable: &'a mut Peekable<Chars<'a>>
+}
+
+fn is_whitespace(c: char) -> bool {
+    matches!(c, ' ' | '\r' | '\t')
+}
+
+impl<'a> Lexer<'a> {
+    pub fn new(
+        input: &'a str,
+        peekable: &'a mut Peekable<Chars<'a>>
+    ) -> Lexer {
+        Lexer {
+            input,
+            position: 0,
+            peekable,
+        }
+    }
+
+    pub fn tokenize(&mut self) -> (Token, Range) {
+        let start = self.position;
+
+        match self.peekable.peek() {
+            None => (Token::EOF, self.mk_range(start)),
+            Some(c) => match c {
+                c if is_whitespace(*c) => {
+                    self.accumulate_while(&is_whitespace);
+                    self.tokenize()
+                }
+                // TODO: Add other types of numbers on it too
+                c if c.is_ascii_digit() => {
+                    let num = self.accumulate_while(&|x| x.is_ascii_digit() || x == '_');
+                    (Token::LInt(num.replace('_', "").parse().unwrap()), self.mk_range(start))
+                }
+                '(' => self.to_single_token(Token::LPar, start),
+                ')' => self.to_single_token(Token::RPar, start),
+                '+' => self.to_single_token(Token::OpPlus, start),
+                _ => unreachable!(),
```

Comment:
You probably already have plans to change this, but this arm will eventually have to handle the error in a way that doesn't trigger a panic.

Reply:
Yes, I just put it as a kind of `TODO` lmao

Reply:
So I think that you must use `todo!()` instead

Reply:
I didn't know about this function, lmao
I would replace for it